File: artifacts/py-api/routes/training.py
import uuid
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from datetime import datetime, timezone
from db import get_db, TrainingSessionRecord, TrainingMessageRecord, AgentMemoryRecord
from agents.training_agent import run_training_agent

router = APIRouter()


@router.get("/sessions")
def get_training_sessions(db: Session = Depends(get_db)):
    sessions = db.query(TrainingSessionRecord).order_by(TrainingSessionRecord.created_at.desc()).all()
    return {
        "sessions": [{
            "id": s.id,
            "trainingType": s.training_type,
            "title": s.title,
            "createdAt": s.created_at.isoformat() if s.created_at else None,
            "messageCount": s.message_count,
            "lastMessageAt": s.last_message_at.isoformat() if s.last_message_at else None,
        } for s in sessions]
    }


@router.post("/sessions")
def create_training_session(body: dict, db: Session = Depends(get_db)):
    training_type = body.get("trainingType", "general")
    title = body.get("title", "Training Session")
    
    valid_types = ["database_schema", "duplicate_definition", "general"]
    if training_type not in valid_types:
        raise HTTPException(400, f"Invalid trainingType. Must be one of: {valid_types}")
    
    session = TrainingSessionRecord(
        id=str(uuid.uuid4()),
        training_type=training_type,
        title=title,
        created_at=datetime.now(timezone.utc),
        message_count=0,
    )
    db.add(session)
    db.commit()
    
    return {
        "id": session.id,
        "trainingType": session.training_type,
        "title": session.title,
        "createdAt": session.created_at.isoformat(),
        "messageCount": 0,
        "lastMessageAt": None,
    }


@router.post("/sessions/{session_id}/messages")
async def send_training_message(session_id: str, body: dict, db: Session = Depends(get_db)):
    session = db.query(TrainingSessionRecord).filter(TrainingSessionRecord.id == session_id).first()
    if not session:
        raise HTTPException(404, "Training session not found")
    
    user_message = body.get("message", "")
    
    # Load DB history for restart-recovery: if LangGraph MemorySaver is empty
    # (e.g., after a server restart), run_training_agent will re-sync from this list.
    history_records = db.query(TrainingMessageRecord).filter(
        TrainingMessageRecord.session_id == session_id
    ).order_by(TrainingMessageRecord.timestamp).all()
    db_history = [{"role": m.role, "content": m.content} for m in history_records]

    result = await run_training_agent(user_message, session.training_type, session_id, db_history)
    
    user_msg = TrainingMessageRecord(
        id=str(uuid.uuid4()),
        session_id=session_id,
        role="user",
        content=user_message,
        timestamp=datetime.now(timezone.utc),
    )
    asst_msg = TrainingMessageRecord(
        id=str(uuid.uuid4()),
        session_id=session_id,
        role="assistant",
        content=result["response"],
        timestamp=datetime.now(timezone.utc),
    )
    db.add(user_msg)
    db.add(asst_msg)
    
    session.message_count = (session.message_count or 0) + 2
    session.last_message_at = datetime.now(timezone.utc)
    
    if result.get("memorySaved") and result.get("memoryKey"):
        existing = db.query(AgentMemoryRecord).filter(
            AgentMemoryRecord.key == result["memoryKey"]
        ).first()
        if existing:
            existing.content = result.get("memoryContent", "")
            existing.updated_at = datetime.now(timezone.utc)
        else:
            memory = AgentMemoryRecord(
                id=str(uuid.uuid4()),
                category=session.training_type,
                key=result["memoryKey"],
                content=result.get("memoryContent", ""),
                created_at=datetime.now(timezone.utc),
                updated_at=datetime.now(timezone.utc),
            )
            db.add(memory)
    
    db.commit()
    
    return {
        "response": result["response"],
        "memorySaved": result.get("memorySaved", False),
        "memoryKey": result.get("memoryKey"),
    }


@router.get("/memory")
def get_agent_memory(db: Session = Depends(get_db)):
    memories = db.query(AgentMemoryRecord).order_by(AgentMemoryRecord.updated_at.desc()).all()
    return {
        "memories": [{
            "id": m.id,
            "category": m.category,
            "key": m.key,
            "content": m.content,
            "createdAt": m.created_at.isoformat() if m.created_at else None,
            "updatedAt": m.updated_at.isoformat() if m.updated_at else None,
        } for m in memories],
        "totalEntries": len(memories),
    }


@router.delete("/memory")
def clear_agent_memory(db: Session = Depends(get_db)):
    deleted = db.query(AgentMemoryRecord).delete()
    db.commit()
    return {"success": True, "message": f"Cleared {deleted} memory entries"}

File: artifacts/api-server/src/py-api/routes/training.ts
import { Router, Request, Response } from "express";
import { query, queryOne, execute } from "../db.js";
import { v4 as uuidv4 } from "uuid";
import { runTrainingAgent } from "../agents/trainingAgent.js";

const router = Router();

router.get("/sessions", async (_req, res) => {
  try {
    const rows = await query("SELECT * FROM dup_training_sessions ORDER BY last_message_at DESC NULLS LAST, created_at DESC");
    res.json({ sessions: rows.map((r) => ({
      id: r["id"],
      trainingType: r["training_type"],
      title: r["title"],
      createdAt: r["created_at"],
      lastMessageAt: r["last_message_at"],
      messageCount: Number(r["message_count"] || 0),
    })) });
  } catch (e) {
    console.error("List training sessions error:", e);
    res.status(500).json({ error: "Failed to list sessions" });
  }
});

router.post("/sessions", async (req, res) => {
  try {
    const { trainingType, title } = req.body as { trainingType: string; title?: string };
    const id = uuidv4();
    await execute(
      `INSERT INTO dup_training_sessions (id, training_type, title, created_at, message_count)
       VALUES ($1, $2, $3, NOW(), 0)`,
      [id, trainingType, title || `${trainingType} Session`]
    );
    const row = await queryOne("SELECT * FROM dup_training_sessions WHERE id = $1", [id]);
    res.json({
      id: row!["id"],
      trainingType: row!["training_type"],
      title: row!["title"],
      createdAt: row!["created_at"],
      lastMessageAt: null,
      messageCount: 0,
    });
  } catch (e) {
    console.error("Create training session error:", e);
    res.status(500).json({ error: "Failed to create session" });
  }
});

router.get("/sessions/:sessionId/messages", async (req, res) => {
  try {
    const rows = await query(
      "SELECT * FROM dup_training_messages WHERE session_id = $1 ORDER BY timestamp ASC",
      [req.params["sessionId"]]
    );
    res.json({ messages: rows.map((r) => ({ id: r["id"], role: r["role"], content: r["content"], timestamp: r["timestamp"] })) });
  } catch (e) {
    console.error("Get training messages error:", e);
    res.status(500).json({ error: "Failed to get messages" });
  }
});

// Shared chat handler — handles both /sessions/:id/messages (schema) and /sessions/:id/chat (legacy)
async function handleTrainingChat(req: Request, res: Response): Promise<void> {
  try {
    const { message } = req.body as { message: string };
    const sessionId = req.params["sessionId"] || req.params["id"];

    const session = await queryOne("SELECT * FROM dup_training_sessions WHERE id = $1", [sessionId]);
    if (!session) { res.status(404).json({ error: "Session not found" }); return; }

    const msgId = uuidv4();
    await execute(
      `INSERT INTO dup_training_messages (id, session_id, role, content, timestamp) VALUES ($1, $2, 'user', $3, NOW())`,
      [msgId, sessionId, message]
    );

    const history = await query(
      "SELECT role, content FROM dup_training_messages WHERE session_id = $1 ORDER BY timestamp ASC",
      [sessionId]
    );

    const { response, memorySaved, memoryKey, memoryContent } = await runTrainingAgent({
      userMessage: message,
      trainingType: String(session["training_type"]),
      sessionId,
      dbHistory: history as Array<{ role: string; content: string }>,
    });

    const replyId = uuidv4();
    await execute(
      `INSERT INTO dup_training_messages (id, session_id, role, content, timestamp) VALUES ($1, $2, 'assistant', $3, NOW())`,
      [replyId, sessionId, response]
    );

    await execute(
      `UPDATE dup_training_sessions SET last_message_at = NOW(), message_count = message_count + 2 WHERE id = $1`,
      [sessionId]
    );

    if (memorySaved && memoryKey && memoryContent) {
      await execute(
        `INSERT INTO dup_agent_memory (id, category, key, content, created_at, updated_at)
         VALUES ($1, $2, $3, $4, NOW(), NOW())
         ON CONFLICT (key) DO UPDATE SET content = EXCLUDED.content, updated_at = NOW()`,
        [uuidv4(), String(session["training_type"]), memoryKey, memoryContent]
      ).catch(() =>
        execute(
          `INSERT INTO dup_agent_memory (id, category, key, content, created_at, updated_at) VALUES ($1, $2, $3, $4, NOW(), NOW())`,
          [uuidv4(), String(session["training_type"]), memoryKey, memoryContent]
        )
      );
    }

    res.json({ response, memorySaved, memoryKey: memoryKey ?? null });
  } catch (e) {
    console.error("Training chat error:", e);
    res.status(500).json({ error: "Chat failed" });
  }
}

// POST /sessions/:id/messages  — matches generated client schema
router.post("/sessions/:sessionId/messages", handleTrainingChat);
// POST /sessions/:id/chat      — legacy alias kept for backward compat
router.post("/sessions/:sessionId/chat", handleTrainingChat);

router.get("/memory", async (_req, res) => {
  try {
    const rows = await query("SELECT * FROM dup_agent_memory ORDER BY updated_at DESC");
    res.json({ memories: rows.map((r) => ({
      id: r["id"],
      category: r["category"],
      key: r["key"],
      content: r["content"],
      createdAt: r["created_at"],
      updatedAt: r["updated_at"],
    })) });
  } catch (e) {
    res.status(500).json({ error: "Failed to get memory" });
  }
});

// DELETE /memory — clear all agent memory entries
router.delete("/memory", async (_req, res) => {
  try {
    await execute("DELETE FROM dup_agent_memory");
    res.json({ success: true, message: "All agent memory cleared" });
  } catch (e) {
    console.error("Clear memory error:", e);
    res.status(500).json({ error: "Failed to clear memory" });
  }
});

export default router;
